_posts/2021-9-28-diff-nil-void.md

```diff
@@ -0,0 +1,41 @@
+---
+layout: post
+title: The difference between nil and "void"
+author: artblart
+---
+
+A common misconception in Lua is that functions have an implicit return value. This is not true. Functions have an implicit return *instruction*, meaning a function will always return after execution even if no return is specified. This implicit return instruction will never carry any value.
+
+This is not to be confused with `nil`, however. For the sake of explanation, we will call this complete absence of value `void`. The source code for Lua 5.1 refers to it as `None` but `void` is a much more common term in the world of programming.
+
+The difference between the two comes from the way emptiness is represented in Lua. `nil` is a value that represents the absence of a value. It is possible to work with `nil` in Lua because it is a native datatype. For the sake of the analogy we will call `nil` a "tangible nothing." It is a value that we can use that represents nothing. `void` on the other hand, is an "intangible nothing." Lua does not comprehend this lack of value because it was designed to use `nil` as a lack of value. The only way to detect this nothingness is through C functions. Take this for example:
+
+```
```

Comment:
The site supports language specific syntax highlighting in the same way it would go into Discord, so please add "lua"